styx-scheduler-service/src/main/java/com/spotify/styx/docker/LabelValue.java

```diff
@@ -0,0 +1,86 @@
+/*-
+ * -\-\-
+ * Spotify Styx Scheduler Service
+ * --
+ * Copyright (C) 2016 - 2020 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.docker;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.regex.Pattern;
+
+final class LabelValue {
+
+  private static final int KUBERNETES_LABEL_MAX_LENGTH = 63;
+  private static final int DIGEST_SUFFIX_LENGTH = 7;
+  private static final int PREFIX_MAX_LENGTH = KUBERNETES_LABEL_MAX_LENGTH - DIGEST_SUFFIX_LENGTH;
+
+  private static final Pattern VALID =
+      Pattern.compile(String.format("^(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9._-]{0,%s}[a-zA-Z0-9])$",
+          KUBERNETES_LABEL_MAX_LENGTH - 2));
+  private static final Pattern INVALID = Pattern.compile("[^a-zA-Z0-9._-]");
```

Comment:
Regex are generally difficult to read, so this is why I like to use constants with descriptive names so I can follow them more easily
```suggestion
  private static final String TERMINAL_CHAR = "[a-zA-Z0-9]";
  private static final String BETWEEN_CHAR = "[a-zA-Z0-9._-]";
  private static final String MANY_BETWEEN_CHAR = String.format(BETWEEN_CHAR + "{0,%s}", KUBERNETES_LABEL_MAX_LENGTH - 2);
  private static final String LONG_LABEL = TERMINAL_CHAR + MANY_BETWEEN_CHAR + TERMINAL_CHAR;
  private static final Pattern VALID =
      Pattern.compile("^(?:" + TERMINAL_CHAR + "|" + LONG_LABEL + ")$");
  private static final Pattern INVALID_CHAR = Pattern.compile("[^a-zA-Z0-9._-]");
```

Reply:
They don't look easier to read for me at all. :(

Reply:
With this I need to either build the complete regex in my head (with a pen to help) or set a breakpoint and run.